source/1_engine/rex_engine/src/cmdline.cpp

```diff
@@ -0,0 +1,205 @@
+#include "rex_engine/cmdline.h"
+
+#include "rex_engine/cmd_args.h"
+#include "rex_engine/diagnostics/assert.h"
+#include "rex_engine/diagnostics/logging/log_macros.h"
+#include "rex_engine/log.h"
+#include "rex_std/cstring.h"
+#include "rex_std/vector.h"
+
+namespace rex
+{
+  namespace cmdline
+  {
+    struct ActiveArgument
+    {
+      rsl::string_view argument;
+      rsl::string_view value;
+    };
+
+    class CommandLineArguments
+    {
+    public:
+      explicit CommandLineArguments(rsl::string_view cmdLine)
+      {
+        // verify the auto generated command line arguments
+        REX_ASSERT_X(verify_args(g_command_line_args.data(), g_command_line_args.size()), "You have ambuguous command line arguments");
+
+        if(cmdLine.empty())
+        {
+          REX_LOG(LogEngine, "No command line arguments to parse");
+          return;
+        }
+
+        parse_cmd_line(rsl::string_view(cmdLine));
+      }
+
+      rsl::optional<rsl::string_view> get_argument(rsl::string_view arg)
+      {
+        for(ActiveArgument active_arg: m_arguments)
+        {
+          // early optimization that strincmp can't do
+          if(arg.length() != active_arg.argument.length())
+            continue;
+
+          if(rsl::strincmp(arg.data(), active_arg.argument.data(), arg.length()) == 0)
+          {
+            return active_arg.value;
+          }
+        }
+
+        return rsl::nullopt;
+      }
+
+    private:
+      void parse_cmd_line(rsl::string_view cmdLine)
+      {
+        // all we have to do is split the command line based with spaces
+        // there's no need to perform any heap allocation for that.
+        count_t start_pos = cmdLine.find_first_not_of(' ', 0); // it's possible, although unlikely, that we start with a space
+
+        // all executables should be called with at least 1 argument
+        // the first argument should always be the path to the executable itself
+        REX_ASSERT_X(start_pos != -1, "No arguments given to your exe, there should always be at least 1 arg.");
+
+        // the module we're launching is always the first argument on the commandline
+        // we skip this
+        count_t space_pos = cmdLine.find_first_of(' ', start_pos);
+
+        // if we have no arguments, we just return
+        // this is possible when running without a debugger
+        if (space_pos == -1)
+        {
+          return;
+        }
+
+        start_pos         = space_pos + 1;
+
+        // get the new space pos
+        space_pos = cmdLine.find_first_of(' ', start_pos);
+
+        const rsl::string_view arg_prefix = "-"; // all arguments should start with a '-'
```

Comment:
What if they don't?

Reply:
then you get an error that your argument isn't found